### ISSUE TITLE
SFTP algorithm settings in <Global> section not being used

### ISSUE THREAD

**Issue by KratiBansal** [proftpd/proftpd#1712]
We have a proftpd server running with mod_sftp to be able to do sftp which is currently using proftpd 1.3.8 and we are planning to remove the support for weak ciphers namely sha1 and blowfish. 

Even after removing the support for sha1 and blowfish, I can see that the server continues to use the deprecated algorithms -
```
Starting Nmap 6.40 ( http://nmap.org ) at 2023-09-06 06:57 UTC
Nmap scan report for ec2-174-129-222-205.compute-1.amazonaws.com (174.xxx.xxx.205)
Host is up (0.0044s latency).
PORT   STATE SERVICE VERSION
22/tcp open  ssh     (protocol 2.0)
1 service unrecognized despite returning data. If you know the service/version, please submit the following fingerprint at http://www.insecure.org/cgi-bin/servicefp-submit.cgi :
SF-Port22-TCP:V=6.40%I=7%D=9/6%Time=64F822D6%P=x86_64-koji-linux-gnu%r(NUL
SF:L,412,"SSH-2\.0-mod_sftp\r\n\0\0\x03\xfc\x06\x14\xba\$\x08\xb3;\xad\n\x
SF:8a\\\xc0#J\xec\xa2\xd7\xbe\0\0\x01\x0eecdh-sha2-nistp521,ecdh-sha2-nist
SF:p384,ecdh-sha2-nistp256,diffie-hellman-group18-sha512,diffie-hellman-gr
SF:oup16-sha512,diffie-hellman-group14-sha256,diffie-hellman-group-exchang
SF:e-sha256,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha1
SF:,rsa1024-sha1,ext-info-s\0\0\0\)rsa-sha2-512,rsa-sha2-256,ssh-rsa,ssh-d
SF:ss\0\0\0\x81aes256-ctr,aes192-ctr,aes128-ctr,aes256-gcm@openssh\.com,ae
SF:s128-gcm@openssh\.com,aes256-cbc,aes192-cbc,aes128-cbc,3des-ctr,3des-cb
SF:c\0\0\0\x81aes256-ctr,aes192-ctr,aes128-ctr,aes256-gcm@openssh\.com,aes
SF:128-gcm@openssh\.com,aes256-cbc,aes192-cbc,aes128-cbc,3des-ctr,3des-cbc
SF:\0\0\0\xa5hmac-sha2-256,hmac-sha2-256-etm@openssh\.com,hmac-sha2-512,hm
SF:ac-sha2-512-etm@openssh\.com,hmac-sha1,hmac-sha1-etm@openssh\.com,hmac-
SF:sha1-96,hmac-sha1-96-etm@openssh\.com\0\0\0\xa5hmac-sha2-256,hmac-sha2-
SF:256-etm@openssh\.com,hmac-sha2-512,hmac-sha2-512-etm@openssh\.com,hmac-
SF:sha1,hmac");

Service detection performed. Please report any incorrect results at http://nmap.org/submit/ .
Nmap done: 1 IP address (1 host up) scanned in 6.13 seconds```
```
we were previously using proftpd-1.3.7d wherein the same config file would work fine, however, just so that we could remove certain SFTPHostKeys, we moved to proftpd-1.3.8 but it is not working as expected. Could you please check and help with the same?

This is our proftpd config file for reference - 

```
# This is the ProFTPD configuration file
#
# See: http://www.proftpd.org/docs/directives/linked/by-name.html

# Server Config - config used for anything outside a <VirtualHost> or <Global> context
# See: http://www.proftpd.org/docs/howto/Vhost.html

ServerName			"ProFTPD server"
ServerIdent			on "FTP Server ready."
ServerAdmin			hosted+devsfs1242fe1@coupa.com
DefaultServer			on

# CEDESK-1839 : Disable default port 21
# http://www.proftpd.org/docs/directives/linked/config_ref_Port.html
Port				0

# COUPA: faster directory listing
AllowOverride                   off

# Don't do reverse DNS lookups (hangs on DNS problems)
UseReverseDNS			off

# Set the user and group that the server runs as
User				nobody
Group				nobody

# To prevent DoS attacks, set the maximum number of child processes
# to 20.  If you need to allow more than 20 concurrent connections
# at once, simply increase this value.  Note that this ONLY works
# in standalone mode; in inetd mode you should use an inetd server
# that allows you to limit maximum number of processes per service
# (such as xinetd)
MaxInstances			2921

# Disable sendfile by default since it breaks displaying the download speeds in
# ftptop and ftpwho
UseSendfile			off

# Define the log formats
LogFormat			default	"%h %l %u %t \"%r\" %s %b"
LogFormat			auth	"%v [%P] %h %t \"%r\" %s"

# Dynamic Shared Object (DSO) loading
# See README.DSO and howto/DSO.html for more details
<IfModule mod_dso.c>
  LoadModule mod_clamav.c
  LoadModule mod_sql.c
  LoadModule mod_sql_passwd.c
  # Load mysql or sqlite
  LoadModule mod_sql_mysql.c
  LoadModule mod_rewrite.c
  LoadModule mod_sftp.c
  LoadModule mod_sftp_sql.c
  LoadModule mod_ctrls_admin.c
  LoadModule mod_ifsession.c
  LoadModule mod_statsd.c
</IfModule>

<IfModule mod_ctrls_admin.c>
  AdminControlsEngine on
  AdminControlsACLs shutdown,status allow user deploy,root
</IfModule>

# Dynamic ban lists (http://www.proftpd.org/docs/contrib/mod_ban.html)
# Enable this with PROFTPD_OPTIONS=-DDYNAMIC_BAN_LISTS in /etc/sysconfig/proftpd
<IfDefine DYNAMIC_BAN_LISTS>
  LoadModule      mod_ban.c
  BanEngine       on
  BanLog          /var/log/proftpd/ban.log
  BanTable        /var/run/proftpd/ban.tab

  # If the same client reaches the MaxLoginAttempts limit 2 times
  # within 10 minutes, automatically add a ban for that client that
  # will expire after one hour.
  BanOnEvent      MaxLoginAttempts 2/00:10:00 01:00:00

  # Allow the FTP admin to manually add/remove bans
  BanControlsACLs   all allow user ftpadm
</IfDefine>


# Global Config - config common to Server Config and all virtual hosts
# See: http://www.proftpd.org/docs/howto/Vhost.html
<Global>

  # Cause every FTP user except adm to be chrooted into their home directory
  # Aliasing /etc/security/pam_env.conf into the chroot allows pam_env to
  # work at session-end time (http://bugzilla.redhat.com/477120)
  #VRootEngine			on
  #VRootAlias			/etc/security/pam_env.conf /etc/security/pam_env.conf

  DefaultRoot			~ !adm

  # Umask 022 is a good standard umask to prevent new dirs and files
  # from being group and world writable
  Umask				022

  # Allow users to overwrite files and change permissions
  AllowOverwrite		yes
  <Limit ALL SITE_CHMOD>
    AllowAll
  </Limit>

  # Allow clients to resume uploads
  AllowStoreRestart on

  # Enables more safe file uploads
  HiddenStores on

  <IfModule mod_sql.c>
    SQLLogFile /var/log/proftpd/mod_sql.log
    # Define a SELECT statement to retrieve users' authorized SSH keys
    #Allow deploy key to authenticate for users
    SQLNamedQuery get-user-authorized-keys SELECT "`key` FROM sftpuserkeys WHERE name='%U' OR name='deploy'"

    # Define a SELECT statement to retrieve hosts' authorized SSH keys
    SQLNamedQuery get-host-authorized-keys SELECT "`key` FROM sftphostkeys WHERE host='%{0}'"

    SQLBackend              mysql
    # The passwords in MySQL are encrypted using CRYPT
    SQLAuthTypes            Crypt OpenSSL Backend
    SQLAuthenticate         users groups


    # used to connect to the database
    # databasename@host database_user user_password
    SQLConnectInfo  sftp@dev1243dbm.int.coupadev.com sftprw vVPpjW8M2FT+

    # Here we tell ProFTPd the names of the database columns in the "usertable"
    # we want it to interact with. Match the names with those in the db
    SQLUserInfo     users userid passwd uid gid homedir shell

    # Here we tell ProFTPd the names of the database columns in the "grouptable"
    # we want it to interact with. Again the names match with those in the db
    SQLGroupInfo    sftpgroups groupname gid members

    # set min UID and GID - otherwise these are 999 each
    SQLMinID        500

    # create a user's home directory on demand if it doesn't exist
    CreateHome on

    # Update count every time user logs in
    SQLLog PASS updatecount
    SQLNamedQuery updatecount UPDATE "login_count=login_count+1, accessed=now() WHERE userid='%u'" users

    # Disable after 90 days
    # SQLUserWhereClause "DATEDIFF(now(), accessed) < 90"

    # Update modified everytime user uploads or deletes a file
    #SQLLog  STOR,DELE modified
    #SQLNamedQuery modified UPDATE "modified=now() WHERE userid='%u'" users

    # User quotas
    # ===========
    #  QuotaEngine on
    #  QuotaDirectoryTally on
    #  QuotaDisplayUnits Mb
    #  QuotaShowQuotas on

    #  SQLNamedQuery get-quota-limit SELECT "name, quota_type, per_session, limit_type, bytes_in_avail, bytes_out_avail, bytes_xfer_avail, files_in_avail, files_out_avail, files_xfer_avail FROM ftpquotalimits WHERE name = '%{0}' AND quota_type = '%{1}'"

    #  SQLNamedQuery get-quota-tally SELECT "name, quota_type, bytes_in_used, bytes_out_used, bytes_xfer_used, files_in_used, files_out_used, files_xfer_used FROM ftpquotatallies WHERE name = '%{0}' AND quota_type = '%{1}'"

    #  SQLNamedQuery update-quota-tally UPDATE "bytes_in_used = bytes_in_used + %{0}, bytes_out_used = bytes_out_used + %{1}, bytes_xfer_used = bytes_xfer_used + %{2}, files_in_used = files_in_used + %{3}, files_out_used = files_out_used + %{4}, files_xfer_used = files_xfer_used + %{5} WHERE name = '%{6}' AND quota_type = '%{7}'" ftpquotatallies

    #  SQLNamedQuery insert-quota-tally INSERT "%{0}, %{1}, %{2}, %{3}, %{4}, %{5}, %{6}, %{7}" ftpquotatallies

    #  QuotaLimitTable sql:/get-quota-limit
    #  QuotaTallyTable sql:/get-quota-tally/update-quota-tally/insert-quota-tally
  </IfModule>

  RootLogin off
  RequireValidShell off

  <IfModule mod_clamav.c>
     ClamAV on
     ClamFailsafe true
     ClamServer 127.0.0.1
     ClamStream on
  </IfModule>

  <IfModule mod_statsd.c>
    StatsdEngine on
    StatsdServer 127.0.0.1:8125 proftpd
  </IfModule>


  # Coupa custom settings
  # Maximum of 10 concurrent connections per account
  # Idle timeout of 2 minutes
  MaxClientsPerUser 1497
  TimeoutIdle 120
  SFTPCiphers  \
    aes128-cbc \
    aes192-cbc \
    aes256-cbc \
    aes128-ctr \
    aes192-ctr \
    aes256-ctr \
    cast128-cbc \
    3des-ctr \
    3des-cbc 
  SFTPDigests \
    hmac-sha2-256  \
    hmac-sha2-512 \
    hmac-ripemd160 \
    umac-64@openssh.com
  SFTPHostKeys  \
    ecdsa-sha2-nistp256  \
    ecdsa-sha2-nistp384  \
    ecdsa-sha2-nistp521  \
    rsa-sha2-512  \
    rsa-sha2-256
  SFTPKeyExchanges \
    diffie-hellman-group-exchange-sha256 \
    ecdh-sha2-nistp256 \
    ecdh-sha2-nistp384 \
    ecdh-sha2-nistp521
</Global>

MaxLoginAttempts 3

# Virtual Hosts are below
#
<VirtualHost 0.0.0.0>
  # Configure the server to listen on the normal SSH2 port, port 22
  Port 8022

  <IfModule mod_sftp.c>
    SFTPEngine on
    SFTPAuthMethods publickey password
    SFTPLog /var/log/proftpd/sftp-port-22.log

    # Configure both the RSA and DSA host keys, using the same host key
    # files that OpenSSH uses.
    SFTPHostKey /etc/proftpd/sftp_host_rsa_key
    SFTPHostKey /etc/proftpd/sftp_host_dsa_key

      FSOptions IgnoreExtendedAttributes

    <IfModule mod_sftp_sql.c>
      # Instead of using a file-based key store, we tell mod_sftp to use
      # the SQL-based key store provided by mod_sftp_sql
      SFTPAuthorizedUserKeys sql:/get-user-authorized-keys
      SFTPAuthorizedHostKeys sql:/get-host-authorized-keys
    </IfModule>

    # Enable compression
    SFTPCompression delayed

    SFTPClientMatch .* channelWindowSize 256MB
  </IfModule>

  # Save on lstat ans stat on .ftpaccess file if they are not used.
  AllowOverride off
</VirtualHost>
```

**Comment by Castaglia** [comment]
From your configuration, you have ProFTPD/mod_sftp configured to listen on port 8022.  I can't tell, from your `nmap` command, whether it is connecting to the ProFTPD server process you expect, or whether there is some other ProFTPD process/configuration being scanned...

**Comment by KratiBansal** [comment]
Hi @Castaglia we are using port 22 for sftp.

**Comment by dmitry-pisarev-coupa** [comment]
There is some logic in iptables NAT chain - e.g.
```
sudo iptables -L -v -n -t nat
...
Chain redirect_to_proftpd_sftp (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     tcp  --  *      *       10.0.11.254          0.0.0.0/0            tcp dpt:22 /* redirect_to_proftpd_sftp */
...
```
Probably, customers use sftp without specifying the port and essentially connect to proftpd, thanks to these redirects.
But I agree that if we run nmap locally, we need to use 8022 port.
@KratiBansal @Castaglia

**Comment by KratiBansal** [comment]
Even without Nmap, I could confirm via rebex ssh check that all the ciphers were being used with 1.3.8. But with the older proftpd version 1.3.7d the expected behaviour was observed.

**Comment by KratiBansal** [comment]
hI @Castaglia could you please provide some steps or suggestions that we can try out for such cases?

**Comment by Castaglia** [comment]
I'm currently working on tracking down the cause of this unexpected behavior.

I have reproduced the reported behavior, even using the latest source code in the `master` branch, when using _e.g._:
```
Port 0

<Global>
  ...
</Global>

<VirtualHost 127.0.0.1>
  Port 2222

  <IfModule mod_sftp.c>
    SFTPEngine on
    ...
  </IfModule>
</VirtualHost>
```

What is interesting, though, is that I get the _expected_ behavior if I _don't_ use a `<VirtualHost>` section, and instead use:
```
Port 2222
DefaultAddress 127.0.0.1

<Global>
  ...
</Global>

<IfModule mod_sftp.c>
  SFTPEngine on
    ...
</IfModule>
```

**Comment by Castaglia** [comment]
In my local testing, I _think_ that the behavior you are seeing is related to Issue #1444.  In particular, some of the changes in https://github.com/proftpd/proftpd/commit/117494e74c8bf775c53da52cb29866d0f7eccd49, specifically when handling the `SFTPCiphers`, `SFTPDigests`, `SFTPKeyExchanges`, and `SFTPHostKeys` directives, lead to this regression.

I noticed, in my local tests, that when I used a barebones configuration, trying to remove everything except what was needed to reproduce this behavior, I ended with something like this:
```
Port 0

<Global>
  SFTPCiphers ...
  SFTPDigests ...
  SFTPKeyExchanges ...
  SFTPHostKeys ...
</Global>

<VirtualHost 127.0.0.1>
  Port 2222

  <IfModule mod_sftp.c>
    SFTPEngine on
    ...
  </IfModule>
</VirtualHost>
```
Notice that my `<Global>` section _only_ had these directives for `mod_sftp`, and nothing else.

When I started ProFTPD manually with debug logging, using `/usr/local/sbin/proftpd -nd10`, I saw the following log message emitted on startup:
```
  proftpd[9190]: </Global>: ignoring empty section
```
Clearly that `<Global>` section was not empty, but the log message said it was.  This pointed me in the direction of the problem.  For whatever reason, the config parser was not "seeing" these SFTP algorithm directives.  And that would indeed explain the behaviors observed.

I'll have a PR shortly which fixes the handling of those directives in `mod_sftp`, reverting some of the changes from that commit -- but also ensuring that I don't regress the functionality for Issue #1444 either!

**Comment by Castaglia** [comment]
https://github.com/proftpd/proftpd/issues/1712 seems to fix this behavior, according to my local tests.  Would you be able to try it out in your setup, verify that it works for you as well?

**Comment by Castaglia** [comment]
I should note that, in the mean time, there is a workaround you should be able to use in your configuration: specify those SFTP algorithms directly in the SFTP `<VirtualHost>` section _as well_:
```
<VirtualHost 0.0.0.0>
  # Configure the server to listen on the normal SSH2 port, port 22
  Port 8022

  <IfModule mod_sftp.c>
    SFTPEngine on
    SFTPAuthMethods publickey password
    SFTPLog /var/log/proftpd/sftp-port-22.log

    # Configure both the RSA and DSA host keys, using the same host key
    # files that OpenSSH uses.
    SFTPHostKey /etc/proftpd/sftp_host_rsa_key
    SFTPHostKey /etc/proftpd/sftp_host_dsa_key

      FSOptions IgnoreExtendedAttributes

    <IfModule mod_sftp_sql.c>
      # Instead of using a file-based key store, we tell mod_sftp to use
      # the SQL-based key store provided by mod_sftp_sql
      SFTPAuthorizedUserKeys sql:/get-user-authorized-keys
      SFTPAuthorizedHostKeys sql:/get-host-authorized-keys
    </IfModule>

    # Enable compression
    SFTPCompression delayed

    SFTPClientMatch .* channelWindowSize 256MB

    # Copied here from our <Global> section due to Issue #1712
    SFTPCiphers  \
      aes128-cbc \
      aes192-cbc \
      aes256-cbc \
      aes128-ctr \
      aes192-ctr \
      aes256-ctr \
      cast128-cbc \
      3des-ctr \
      3des-cbc 
    SFTPDigests \
      hmac-sha2-256  \
      hmac-sha2-512 \
      hmac-ripemd160 \
      umac-64@openssh.com
    SFTPHostKeys  \
      ecdsa-sha2-nistp256  \
      ecdsa-sha2-nistp384  \
      ecdsa-sha2-nistp521  \
      rsa-sha2-512  \
      rsa-sha2-256
    SFTPKeyExchanges \
      diffie-hellman-group-exchange-sha256 \
      ecdh-sha2-nistp256 \
      ecdh-sha2-nistp384 \
      ecdh-sha2-nistp521
  </IfModule>

  # Save on lstat ans stat on .ftpaccess file if they are not used.
  AllowOverride off
</VirtualHost>
```

**Comment by Castaglia** [comment]
I've merged the fix to the `master` branch, along with a regression test, and backported the fix to the 1.3.8 branch.  Thanks for reporting this issue!

**Comment by KratiBansal** [comment]
Thanks for fixing it @Castaglia. I will try building the package with the latest changes and let you know.
